booklog/src/App.tsx

```diff
@@ -4,55 +4,58 @@ import { BookItem } from './types/index' //この型データ使うよ
-  const randomBook = bookData[randomBookIndex]; //randomBook変数に格納する
-
+  const [bookData, setBookData] = useState<BookItem[]>(MOCK_DATA.items); //MOCK_DATAのitemsからランダムにデータを取得し、BookItemの型配列にならって保持するstate。
+  console.log(bookData);
   return (
     <>
-      <main>
-        {
-          randomBook && (
-            <div className="books-detail">
-              {
-                randomBook.volumeInfo.imageLinks && (
-                  <img src={randomBook.volumeInfo.imageLinks.thumbnail} className="books-img" alt={randomBook.volumeInfo.title} width="128" height="169" />
-                )
-              }
-              <div className="books-detail-contents">
-                <h2 className="book-ttl">{randomBook.volumeInfo.title}</h2>
-                {
-                  randomBook.volumeInfo.description && (
-                    <p className="book-description">
-                      {randomBook.volumeInfo.description}
-                    </p>
-                  )
-                }
-                {
-                  randomBook.volumeInfo.authors && (
-                    <p className="book-authors-list">著者：
-                      {
-                        randomBook.volumeInfo.authors.map((author) => (
-                          <span className="author" key={randomBook.id}>{author}</span>
-                        ))
-                      }
-                    </p>
-                  )
-                }
-                {
-                  randomBook.volumeInfo.publisher && (
-                    <p className="book-publisher">出版社：{randomBook.volumeInfo.publisher}</p>
-                  )
-                }
+      <header className="header">
+        <h1>Booklog</h1>
+      </header>
+      <main className="main">
+        <div className="books-box">
+          {
+            bookData.map(({ id, volumeInfo: { imageLinks, title, description, authors, publisher, previewLink } }) => (
+              <div key={id} className="books-detail">
                 {
-                  randomBook.volumeInfo.previewLink && (
-                    <a href={randomBook.volumeInfo.previewLink} className="book-link">詳しく見る</a>
+                  imageLinks && (
+                    <img src={imageLinks.thumbnail} className="books-img" alt={title} width="128" height="169" />
                   )
                 }
+                <div className="books-detail-contents">
+                  <h2 className="book-ttl">{title}</h2>
+                  {
+                    description && (
+                      <p className="book-description">
+                        {description}
+                      </p>
+                    )
+                  }
+                  {
+                    authors && (
+                      <p className="book-authors-list">著者：
+                        {
+                          authors.map((author) => (
+                            <span className="author" key={id}>{author}</span>
```

Comment:
🚨ここは注意ポイントです！！🚨

chrome dev toolsのconsoleを確認すると以下のようなエラーが出ています。
> Encountered two children with the same key, `SfUqzwEACAAJ`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.

これは同じキーを持つ要素が2つ以上存在することによるエラーです。
複数の`<span class="author">`間で、同じ文字列を`key`属性に指定していることが原因で発生しています。

では`key`属性の重要性及びその属性に指定すべき値に関して、公式ドキュメントの[この箇所](https://ja.react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key)を参照してみましょう。
このドキュメントによると`key`属性に関して以下のような記載があります。
> 配列内の他のアイテムと区別できるようにするための一意な文字列ないし数値のことです。

つまり、各`<span class="author">`に対してそれぞれ別の文字列ないし数値を指定できれば、エラーが解消できるのではないでしょうか。
上記のドキュメントは非常に充実しているため、是非読み込んでいただければ解決のヒントが見つかると思います💡

また最後にはなりますが、consoleにエラーが出力されていないかを確認する癖をつけるとより効率的に実装ができるようになります。心がけてみてください。

Reply:
@NBmaeda 
コメントありがとうございます🙇‍♀️

こちらindexを利用してconsoleエラーの解消を試みたのですが、載せてくださっているドキュメントの中のこの部分であまりindexを利用するのは推奨しないとのコメントがあります👀
https://ja.react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key:~:text=%E3%81%A7%E3%81%8D%E3%82%8B%E3%81%AE%E3%81%A7%E3%81%99%E3%80%82-,%E8%90%BD%E3%81%A8%E3%81%97%E7%A9%B4,-%E3%82%A2%E3%82%A4%E3%83%86%E3%83%A0%E3%81%AE%E3%82%A4%E3%83%B3%E3%83%87%E3%83%83%E3%82%AF%E3%82%B9

今回のような場合、authorの中のデータとしては著者名のみを複数所持、もしくはゼロという可能性があるかと思うのですが、その場合indexを使用することは好ましくないのでしょうか？😥

Reply:
@NByamane 
良い疑問で素晴らしいと思います👏

結論から言うと、このケースでは`author`を`key`属性に使用する形で良いと思います。
上記ドキュメントに「`key`のルール」として以下の記載があります。
>キーは兄弟間で一意でなければなりません。ただし、異なる配列に対応する JSX ノードには同じキーを使用することができます。

著者名の項目の性質を考えると、ある書籍`A`の著者`a`は別の書籍`B`の著者である可能性はありますが、`A`に関してのデータ内に重複して列挙されていることはないと予想できます。
そのため上記のルールに合致しており、使用しても問題ないと判断できます💡

`key`に`index`を使用することが問題視されるのは、配列内の要素が動的に変更されるケースです。
そのため今回の場合大きな問題ではないのですが、可能であれば避けた方が良いです。
[こちら](https://zenn.dev/luvmini511/articles/f7b22d93e9c182)の記事にこの問題をわかりやすく体験できるdemoが紹介されているので、余裕があれば是非みてみてください！